### ISSUE TITLE
Query: EntityEqualityRewritingExpressionVisitor doesn't work with object.Equals method

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFrameworkCore#8366]
```cs
from c1 in ctx.Customers
from c2 in ctx.Customers
where c1.CustomerID == "ALFKI"
where c1 == c2
select c1.CustomerID
```

works fine (returns 1 result), however:

```cs
from c1 in ctx.Customers
from c2 in ctx.Customers
where c1.CustomerID == "ALFKI"
where c1.Equals(c2)
select c1.CustomerID
```

doesn't return any results, basically we perform client eval here and do reference comparison between c1 and c2. This works ok for collections (apart from the predicate being executed on the client), but returns unexpected results if we compare collections:

```cs
from c1 in ctx.Customers
from c2 in ctx.Customers
where c1.CustomerID == "ALFKI"
where c1Orders.Equals(c2.Orders)
select c1.CustomerID
```

 The query above returns no results.

We should handle this case also, for the most part should be a matter of translating it to c1 == c2 and revisiting. Just need to detect a case when types are not matching - in case of binary expression, compiler handles that for us, but Equals method is loosely typed

**Comment by smitpatel** [comment]
Interesting, shouldn't c1 & c2 should resolve to single tracked reference in the context?

**Comment by maumar** [comment]
This is probably because we materialize both collections with MaterializeCollectionNavigation, creating new instances every time. Here is the query plan:

```
(QueryContext queryContext) => IEnumerable<<>f__AnonymousType68<string, string>> _InterceptExceptions(
    source: IEnumerable<<>f__AnonymousType68<string, string>> _Select(
        source: IEnumerable<TransparentIdentifier<ValueBuffer, ValueBuffer>> _Where(
            source: IEnumerable<TransparentIdentifier<ValueBuffer, ValueBuffer>> _ShapedQuery(
                queryContext: queryContext, 
                shaperCommandContext: SelectExpression: 
                    SELECT [c1].[CustomerID] AS [Id1], [c2].[CustomerID] AS [Id2]
                    FROM [Customers] AS [c1]
                    CROSS JOIN [Customers] AS [c2]
                    WHERE ([c1].[CustomerID] = N'ALFKI') AND ([c2].[CustomerID] = N'ALFKI'), 
                shaper: TypedCompositeShaper<ValueBufferShaper, ValueBuffer, ValueBufferShaper, ValueBuffer, TransparentIdentifier<ValueBuffer, ValueBuffer>>), 
            predicate: (TransparentIdentifier<ValueBuffer, ValueBuffer> t0) => bool ICollection<Order> MaterializeCollectionNavigation(
                navigation: Customer.Orders, 
                elements: IEnumerable<Order> _InjectParameters(
                    queryContext: queryContext, 
                    source: IEnumerable<Order> _ShapedQuery(
                        queryContext: queryContext, 
                        shaperCommandContext: SelectExpression: 
                            SELECT [o].[OrderID], [o].[CustomerID], [o].[EmployeeID], [o].[OrderDate]
                            FROM [Orders] AS [o]
                            WHERE @_outer_CustomerID = [o].[CustomerID], 
                        shaper: BufferedEntityShaper<Order>), 
                    parameterNames: new string[]{ "_outer_CustomerID" }, 
                    parameterValues: new object[]{ string TryReadValue(t0.Outer, 0, Customer.CustomerID) })).Equals(ICollection<Order> MaterializeCollectionNavigation(
                    navigation: Customer.Orders, 
                    elements: IEnumerable<Order> _InjectParameters(
                        queryContext: queryContext, 
                        source: IEnumerable<Order> _ShapedQuery(
                            queryContext: queryContext, 
                            shaperCommandContext: SelectExpression: 
                                SELECT [o0].[OrderID], [o0].[CustomerID], [o0].[EmployeeID], [o0].[OrderDate]
                                FROM [Orders] AS [o0]
                                WHERE @_outer_CustomerID1 = [o0].[CustomerID], 
                            shaper: BufferedEntityShaper<Order>), 
                        parameterNames: new string[]{ "_outer_CustomerID1" }, 
                        parameterValues: new object[]{ string TryReadValue(t0.Inner, 1, Customer.CustomerID) })))), 
        selector: (TransparentIdentifier<ValueBuffer, ValueBuffer> t0) => new <>f__AnonymousType68<string, string>(
            string TryReadValue(t0.Outer, 0, Customer.CustomerID), 
            string TryReadValue(t0.Inner, 1, Customer.CustomerID)
        )), 
    contextType: Specification.Tests.TestModels.Northwind.NorthwindContext, 
    logger: DiagnosticsLogger<Query>, 
    queryContext: queryContext)
```

@anpete @divega - thoughts?

**Comment by anpete** [comment]
Looks like same issue as #7569. MaterializeCollectionNavigation is by-passing everything 🐛

**Comment by maumar** [comment]
Fixed in c4129dc2bef433f5ce55a21ec7a417ba8f153752

**Comment by smitpatel** [comment]
The commit it references actually marks the test disabled with this issue number.
